meta/client.go

```diff
@@ -0,0 +1,216 @@
+
+type ClusterManager struct {
+	Mut    sync.RWMutex
+	ZkConn *zk.Conn
+	// table->TableInfoWatcher
+	Tables gcache.Cache
+	// metaAddrs->metaManager
+	Metas map[string]*session.MetaManager
+}
+
+// zkContext cancels the goroutine that's watching the zkNode, when the watcher
+// is evicted by LRU.
+type zkContext struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+type TableInfoWatcher struct {
+	tableName   string
+	clusterName string
+	metaAddrs   string
+	event       <-chan zk.Event
+	ctx         zkContext
+}
+
+// TODO(jishuo1) change log module
+func initClusterManager() {
+	zkConn, _, err := zk.Connect(zkAddrs, time.Duration(zkTimeOut))
+	if err != nil {
+		logrus.Panicf("failed to connect to zookeeper \"%s\": %s", zkAddrs, err)
+	}
+
+	tables := gcache.New(zkWatcherCount).LRU().EvictedFunc(func(key interface{}, value interface{}) {
+		value.(*TableInfoWatcher).ctx.cancel()
+		logrus.Warnf("table [%s] evicted from table cache (capacity: %d)", key.(string), zkWatcherCount)
+	}).Build() // TODO(jiashuo1) consider set expire time
+	globalClusterManager = &ClusterManager{
+		ZkConn: zkConn,
+		Tables: tables,
+		Metas:  make(map[string]*session.MetaManager),
+	}
+}
+
+func (m *ClusterManager) getMeta(table string) (*session.MetaManager, error) {
+	var meta *session.MetaManager
+
+	tableInfo, err := globalClusterManager.Tables.Get(table)
+	if err == nil {
+		meta = globalClusterManager.Metas[tableInfo.(*TableInfoWatcher).metaAddrs]
+		if meta != nil {
+			return meta, nil
+		}
+	}
+
+	logrus.Printf("[%s] can't get cluster info from local cache, try fetch from zk.", table)
+	globalClusterManager.Mut.Lock()
+	defer globalClusterManager.Mut.Unlock()
+	tableInfo, err = globalClusterManager.Tables.Get(table)
+	if err != nil {
+		tableInfo, err = m.getTableInfo(table)
+		if err != nil {
+			logrus.Errorf("get table[%s] info failed: %s", table, err)
+			return nil, err
+		}
+		err = globalClusterManager.Tables.Set(table, tableInfo)
+		if err != nil {
+			logrus.Warnf("[%s] cluster info update local cache failed: %s", table, err)
+		}
+	}
+
+	metaAddrs := tableInfo.(*TableInfoWatcher).metaAddrs
+	meta = globalClusterManager.Metas[metaAddrs]
+	if meta == nil {
+		metaList, err := parseToMetaList(metaAddrs)
+		if err != nil {
+			logrus.Errorf("The table[%s] cluster addr[%s] format is err: %s", table, metaAddrs, err)
+			return nil, base.ERR_INVALID_DATA
+		}
+		meta = session.NewMetaManager(metaList, session.NewNodeSession)
+		globalClusterManager.Metas[metaAddrs] = meta
+	}
+
+	return meta, nil
+}
+
+// get table cluster info and watch it based table name from zk
+// The zookeeper path layout:
+// /<RegionPathRoot>/<table> =>
+//                         {
+//                           "cluster_name" : "clusterName",
+//                           "meta_addrs" : "metaAddr1,metaAddr2,metaAddr3"
+//                         }
+func (m *ClusterManager) getTableInfo(table string) (*TableInfoWatcher, error) {
```

Comment:
```suggestion
func newTableInfo(table string) (*TableInfoWatcher, error) {
```

you use `globalClusterManager` below (L119). so don't make it a member function of `ClusterManager`